docs/proposals/push-metrics.md

```diff
@@ -0,0 +1,221 @@
+
+Like the existing custom metrics pull mechanism, the metrics pushed through the
+push mechanism are intended to be those metrics that are useful for system
+components, such as metrics inteded for use with autoscaling.  The custom
+metrics pushed through the push mechanisms are still intended to follow the
+overall guidelines for Heapster custom metrics (keep the number of different
+metric names relatively limitted, etc).
+
+Unlike the existing custom metrics mechanism, the push mechanisms are mainly
+intended for use by add-on "cluster-level" components -- components deployed by
+administrators (whether they are running as pods on the cluster, or alongside
+the cluster).  The push mechanisms are not, in general, intended for use by
+arbitrary cluster users (although the metrics transfered would probably apply
+to the users' pods).  It is up to cluster admins to decide which applications
+are permitted to use the push mechanims (see the "Authentication" section below
+for more information on how cluster admins can control access).
+
+For these reasons, the push metrics mechanism is not designed to replace the
+existing pull mechanism for custom metrics, but instead to support additional
+use cases not already supported.
+
+## API
+
+### Authentication, Segregation, and Flow Control
+
+Producers will be authenticated similarly to the way Heapster currently
+authenticates clients: producers will present a certificate signed by a CA, and
+Heapster will be configurable to only allow certain names to push metrics, or
+to allow any certificate signed by the CA.
+
+Additionally, the name presented during authentication will be used as a prefix
+for all metrics added.  This will prevent two different metrics producers from
+accidentally overwriting each other's custom metrics (otherwise, push metrics
+will be stored and retrieved identially to pull-based custom metrics).
+
+In order to prevent push metrics from overwhelming the Heapster instance, it
+will be possible to limit the total number of custom metrics each producer is
+allowed to add, and the frequency at which producers are allowed to push new
+sets of custom metrics.  By default, no limits will be enforced unless
+explicitly set via command line arguments.
+
+### Paths
+
+To add metrics, metrics producers will `POST` new metrics to
+`/api/v1/push/{format}` in the format specified by `{format}`.  The metrics are
+pushed in bulk -- it is up to the format to support a way to name specific
+metrics, namespaces, pods, and containers (for instance, the Prometheus format
+uses metric names and labels for this purpose).
+
+### Metrics Format
+
+The underlying design can support multiple format "backends".  The initial
+backend, detailed here, will be based on the Prometheus format, and will be
+available at `/api/v1/push/prometheus`.
+
+Initially, only the Prometheus text format will be supported. In the future, we
+may also wish to support the promethus protobuf format as well.
```

Comment:
As I've indicated before what you're proposing here is not compatible with the Prometheus ecosystem. You're effectively asking users to hand-produce the Prometheus text format. I don't think is a good idea for either the Kubernetes or Prometheus ecosystems, as it will frustrate everyone involved.

If you want to use our format you should be compatible with our Pushgateways URL pattern, and support both text and protobuf formats.